Kafka quota backpressure

HonoreDB opened the pull request:
Draftiest of draft PRs -- needs tests and logging cleanup, and also I'd like to try to rework just about every aspect of the implementation. But! This PR converts [Kafka throttling](https://docs.confluent.io/kafka/design/quotas.html#request-rate-quotas) protocol messages into changefeed backpressure. See top two commits for details.

adityamaru committed:
jobs: update jobs and job_info join queries

This change updates the two places where we perform
a join over the `system.jobs` and `system.job_info`
table to use more performant and correct queries.

The queries are more performant because of an additional
predicate on the `job_id` in each of the CTE queries. This
prevents a full table scan when we are only concerned with
rows corresponding to a particular job. This has been verified
by running both variants of the query through the EXPLAIN ANALYZE (plan)
tool. This tool occassionally reports a full table scan in the CTE queries
when there is no predicate involved.

The queries are more correct because they take into consideration
a job having >1 row for a particular info_key. This is possible
during upgrades. Jobs start writing their payload and progress to
the job_info table once V23_1CreateSystemJobInfoTable is active.
If a job is still running when the cluster steps through
V23_1JobInfoTableIsBackfilled then the job will have duplicate entries
for the payload and prgress rows. This is because we blindly backfill
the contents of the jobs table into the job_info table without
clearing previous entries corresponding to the job.

This change also adds a red-green regression test for the above
correctness problem.

Fixes: #99101
Release note: None

HonoreDB committed:
changefeedccl: honor Kafka throttling

This commit is a thoroughly unsatisfactory way
of handling the fact that Sarama ignores the
field in responses indicating that we're being
throttled. It's unsatisfactory in that it does
a fragile hack to get into Sarama internals well
enough to know about the response, and in that
all we do is honor the letter of it without
changing our overall byte rate. But it does reduce
the risk of thrashing due to i/o closed errors.

Release note: None

HonoreDB committed:
changefeedccl: add byte-rate throttling to kafka sink

This commit causes each kafkaSink instance to try to
infer what byte rate it should emit at in order to honor
the client-level quota messages it gets from the broker,
and enforce it using a cdcutils.Throttler.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100348)
<!-- Reviewable:end -->

HonoreDB committed:
cdc: use sarama's built in metrics for outgoing byte rate

Previous commit rolled its own byte rate sampler, which is
unnecessary since Sarama has its own that we can yoink a
reference to at sink creation time.

HonoreDB committed:
changefeedccl: adaptive throttling

Sarama and Kafka don't seem to quite agree on what the byte rate is,
even when running in a single node cluster, so if we get repeatedly
throttled, gradually increase a fudge factor until they match.

Release note: None

HonoreDB commented:
Closing in favor of https://github.com/cockroachdb/cockroach/pull/103226
